File: Assets/Code/Logic/Formation/Model/FormationData.cs
using System.Collections.Generic;
using Common.Util;
using Logic.Game.Model;
using Logic.Enums;

namespace Logic.Formation.Model
{
    public class FormationData
    {
        private static Dictionary<int, FormationData> _formationDataDictionary;
        public static Dictionary<int, FormationData> FormationDataDictionary
        {
            get
            {
                if (_formationDataDictionary == null)
                {
                    _formationDataDictionary = CSVUtil.Parse<int, FormationData>("config/csv/formation", "id");
                }
                return _formationDataDictionary;
            }
        }

        public static Dictionary<int, FormationData> GetFormationDatas() 
        {
            if (_formationDataDictionary == null)
            {
                _formationDataDictionary = CSVUtil.Parse<int, FormationData>("config/csv/formation", "id");
            }
            return _formationDataDictionary;
        }

        public static List<FormationData> GetAllFormationDataList()
        {
            return FormationDataDictionary.GetValues();
        }

        public static FormationData GetFormationData(int id)
        {
            FormationData formationData = null;
            FormationDataDictionary.TryGetValue(id, out formationData);
            return formationData;
        }

        public List<FormationPosition> GetAllEnabledPos()
        {
            List<FormationPosition> posList = new List<FormationPosition>();
            for (int i = 0, count = pos.Length; i < count; i++)
            {
                if (pos[i])
                {
                    posList.Add((FormationPosition)(i + 1));
                }
            }
            return posList;
        }

        public bool GetPosEnable(int index)
        {
            if (index < 0 || index >= pos.Length)
                return false;
            return pos[index];
        }
        public bool GetPosEnalbe(FormationPosition position)
        {
            int index = (int)position - 1;
            return GetPosEnable(index);
        }

        [CSVElement("id")]
        public int id;

        [CSVElement("name")]
        public string name;

        [CSVElement("des")]
        public string description;

        public bool[] pos = new bool[9];

        [CSVElement("p1")]
        public int p1
        {
            set
            {
                pos[0] = value != 0;
            }
        }

        [CSVElement("p2")]
        public int p2
        {
            set
            {
                pos[1] = value != 0;
            }
        }

        [CSVElement("p3")]
        public int p3
        {
            set
            {
                pos[2] = value != 0;
            }
        }

        [CSVElement("p4")]
        public int p4
        {
            set
            {
                pos[3] = value != 0;
            }
        }

        [CSVElement("p5")]
        public int p5
        {
            set
            {
                pos[4] = value != 0;
            }
        }

        [CSVElement("p6")]
        public int p6
        {
            set
            {
                pos[5] = value != 0;
            }
        }

        [CSVElement("p7")]
        public int p7
        {
            set
            {
                pos[6] = value != 0;
            }
        }

        [CSVElement("p8")]
        public int p8
        {
            set
            {
                pos[7] = value != 0;
            }
        }

        [CSVElement("p9")]
        public int p9
        {
            set
            {
                pos[8] = value != 0;
            }
        }

        [CSVElement("max_lv")]
        public int maxLevel;

        public GameResData upgradeCostGameResData;
        [CSVElement("upgrade_cost")]
        public string upgradeCostStr
        {
            set
            {
                upgradeCostGameResData = new GameResData(value);
            }
        }
        [CSVElement("upgrade_formula_a")]
        public float upgradeA;

        [CSVElement("upgrade_formula_b")]
        public float upgradeB;

        [CSVElement("upgrade_formula_c")]
        public float upgradeC;

        [CSVElement("upgrade_base_cost_a")]
        public float upgrade_base_a;

        [CSVElement("upgrade_base_cost_b")]
        public float upgrade_base_b;

        public int[] formationConditionIds;
        [CSVElement("formation_condition")]
        public string formationConditionString
        {
            set
            {
                formationConditionIds = value.ToArray<int>(CSVUtil.SYMBOL_SEMICOLON);
            }
        }

        public GameResData costGameResData;
        [CSVElement("unlock_cost")]
        public string cost
        {
            set
            {
                costGameResData = new GameResData(value);
            }
        }
    }
}

File: Assets/ToLua/Lua/user/ui/formation/model/formation_data.lua
local t = {}
t.data = {}
local item = {}
item.__index = item

local game_res_data = require('ui/game/model/game_res_data')

function item.New(table)
	local o = {}
	setmetatable(o,item)
  
	o.id = tonumber(table.id)
	o.name = table.name
	o.des = table.des
  o.pos = {}
	o.pos[1] = tonumber(table.p1) ~= 0
	o.pos[2] = tonumber(table.p2) ~= 0
	o.pos[3] = tonumber(table.p3) ~= 0
	o.pos[4] = tonumber(table.p4) ~= 0
	o.pos[5] = tonumber(table.p5) ~= 0
	o.pos[6] = tonumber(table.p6) ~= 0
	o.pos[7] = tonumber(table.p7) ~= 0
	o.pos[8] = tonumber(table.p8) ~= 0
	o.pos[9] = tonumber(table.p9) ~= 0
	o.max_lv = tonumber(table.max_lv)
	o.upgrade_cost = game_res_data.NewByString(table.upgrade_cost)
	o.upgrade_formula_a = tonumber(table.upgrade_formula_a)
	o.upgrade_formula_b = tonumber(table.upgrade_formula_b)
	o.upgrade_formula_c = tonumber(table.upgrade_formula_c)
	o.upgrade_base_cost_a = tonumber(table.upgrade_base_cost_a)
	o.upgrade_base_cost_b = tonumber(table.upgrade_base_cost_b)
  o.formationConditionIdsList = {}
	local condition = string.split(table.formation_condition,';')
  for k, v in pairs(condition) do
    o.formationConditionIdsList[k] = tonumber(v)
  end
  o.unlock_cost = game_res_data.NewByString(table.unlock_cost)
	return o
end

function item:GetPosEnable(index)
  if index<= 0 or index > #self.pos then 
    return false
  end
  return self.pos[index]
  
end
function item:GetAllEnabledPosList()
  local p = {}
  local index = 1
  for k,v in pairs(self.pos) do
    if v then
      p[index] = k
      index = index + 1
    end
  end
  return p
end

function t.GetDataById(id)
	return t.data[id]
end


local function Start()
	local origin = dofile('formation')
	origin.ForEach(function(id,table)
		local newItem = item.New(table)
		t.data[id] = newItem
	end)
end

Start()
return t